galley/pkg/config/source/kube/apiserver/source.go

```diff
@@ -79,6 +99,106 @@ func (s *Source) Start() {
 	}
 	s.started = true
 
+	// Create a set of pending resources. These will be matched up with incoming CRD events for creating watchers for
+	// each resource that we expect.
+	s.expectedResources = make(map[string]schema.KubeResource)
+	for _, r := range s.options.Resources {
+		// If we received the metadata with the resource marked as disabled, simply ignore it.
+		if r.Disabled {
+			continue
+		}
+
+		// Use the disabled bit to track whether we received an event for this resource.
+		r.Disabled = true
+		s.expectedResources[asKey(r.Group, r.Kind)] = r
+	}
+
+	// Start the CRD listener. When the listener is fully-synced, the listening of actual resources will start.
+	scope.Source.Infof("Beginning CRD Discovery, to figure out resources that are available...")
+	s.provider = rt.NewProvider(s.options.Client, s.options.ResyncPeriod)
+	a := s.provider.GetAdapter(crdKubeResource)
+	s.crdWatcher = newWatcher(crdKubeResource, a)
+	s.crdWatcher.dispatch(event.HandlerFromFn(s.onCrdEvent))
+	s.crdWatcher.start()
+}
+
+func (s *Source) onCrdEvent(e event.Event) {
+	scope.Source.Debuga("onCrdEvent: ", e)
+
+	s.mu.Lock()
```

Comment:
Any concerns about locking for each event? I wonder if we could just always handle incoming events and pass them through to the handlers without locking. This would require the shutdown code to be more sophisticated, but may yield higher throughput.

Reply:
Not particularly. CRD events should be relatively small at the beginning (on the order of 10s) and pretty much non-existent after FullSync. This is a setup/meta level operation, so gains from lock contention avoidance is not significant. Besides, almost all of the events would require locking anyways.